### PR TITLE
Replace `GetXXX()` functions with `AsXXX()`

### PR DESCRIPTION
**Pull request by jmalloc**
This PR modifies the existing changes made for #33 to focus more on providing convenience functions that don't require the user to perform error checking.

To that end, the variants of `GetXXX()` have been removed, and the `MustGetXXX()` functions have been renamed to `AsXXX()` (so as to better match the existing naming of the methods on the `Bucket` interface).

Any variants that returned an `ok` value now panic if the key is undefined.

I've also added `P` variants of the `GetIntNN()` functions that demand a non-zero, positive value, as that is often required even when you need a signed type.

### CODE REVIEW

File: config/int.go [hunk 1]
```diff
@@ -5,92 +5,215 @@ import (
...
+// AsInt16PDefault returns the int16 representation of the value associated with
+// k, or the default value v if k is undefined.
 //
-// It panics if k is defined but its value cannot be parsed as an int.
-func MustGetInt(b Bucket, k string) (v int, ok bool) {
-	v, ok, err := GetInt(b, k)
-	if err != nil {
-		panic(err)
-	}
+// It panics if the value is not zero or negative.
```
Comment (danilvpetrov) [inline config/int.go:87]:
```suggestion
// It panics if the value is not zero or positive.
```

File: config/bytes_test.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+var _ = Describe("func AsBytesDefault()", func() {
+	It("returns a byte-slice value", func() {
+		b := Map{"<key>": String("<value>")}
+
+		v := AsBytesDefault(b, "<key>", []byte("<default>"))
+		Expect(v).To(Equal([]byte("<value>")))
+	})
+
+	It("returns the default value key is not defined", func() {
```
Comment (koden-km) [inline config/bytes_test.go:58]:
```suggestion
	It("returns the default value if key is not defined", func() {
```

File: config/int.go [hunk 3]
```diff
@@ -5,92 +5,215 @@ import (
...
+	return int8(asInt(b, k, 8, true))
+}
 
-	if ok {
-		return x, nil
-	}
+// AsInt8PDefault returns the int8 representation of the value associated with
+// k, or the default value v if k is undefined.
+//
+// It panics if the value is not zero or negative.
```
Comment (danilvpetrov) [inline config/int.go:59]:
```suggestion
// It panics if the value is not zero or positive.
```

File: config/int.go [hunk 4]
```diff
@@ -5,92 +5,215 @@ import (
...
+// panics if unable to do so.
 //
-// If k is defined but its value cannot be parsed as an int, it returns an
-// error describing the invalid value.
-func GetIntDefault(b Bucket, k string, v int) (int, error) {
-	x, ok, err := GetInt(b, k)
-	if err != nil {
-		return 0, err
-	}
+// It panics if the value is not zero or negative.
```
Comment (danilvpetrov) [inline config/int.go:51]:
```suggestion
// It panics if the value is not zero or positive.
```

File: config/int.go [hunk 5]
```diff
@@ -5,92 +5,215 @@ import (
...
+func AsInt16Default(b Bucket, k string, v int16) int16 {
+	return int16(asIntDefault(b, k, 16, int64(v), false))
 }
 
-// MustGetInt returns the int representation of the value associated with k.
+// AsInt16P returns the int16 representation of the value associated with k or
+// panics if unable to do so.
 //
-// If k is undefined, ok is false.
+// It panics if the value is not zero or negative.
```
Comment (danilvpetrov) [inline config/int.go:79]:
```suggestion
// It panics if the value is not zero or positive.
```

File: config/int.go [hunk 6]
```diff
@@ -5,92 +5,215 @@ import (
...
+// or the default value v if k is undefined.
+func AsIntDefault(b Bucket, k string, v int) int {
+	return int(asIntDefault(b, k, 0, int64(v), false))
+}
+
+// AsIntP returns the int representation of the value associated with k or panics
+// if unable to do so.
 //
-// If k is undefined, ok is false and err is nil.
+// It panics if the value is not zero or negative.
```
Comment (koden-km) [inline config/int.go:23]: I think this should not have the *not*?
```suggestion
// It panics if the value is zero or negative.
```

File: config/int.go [hunk 7]
```diff
@@ -5,92 +5,215 @@ import (
...
+// AsInt32Default returns the int32 representation of the value associated with k,
+// or the default value v if k is undefined.
+func AsInt32Default(b Bucket, k string, v int32) int32 {
+	return int32(asIntDefault(b, k, 32, int64(v), false))
+}
+
+// AsInt32P returns the int32 representation of the value associated with k or
+// panics if unable to do so.
+//
+// It panics if the value is not zero or negative.
```
Comment (danilvpetrov) [inline config/int.go]:
```suggestion
// It panics if the value is not zero or positive.
```

File: config/int.go [hunk 8]
```diff
@@ -5,92 +5,215 @@ import (
...
+
+// AsInt32PDefault returns the int32 representation of the value associated with
 // k, or the default value v if k is undefined.
 //
-// It panics if k is defined but its value cannot be parsed as an int.
-func MustGetIntDefault(b Bucket, k string, v int) int {
-	if x, ok := MustGetInt(b, k); ok {
-		return x
-	}
+// It panics if the value is not zero or negative.
```
Comment (danilvpetrov) [inline config/int.go]:
```suggestion
// It panics if the value is not zero or positive.
```

File: config/int.go (hunk 6 above)
Comment (koden-km) [reply config/int.go]: Or flip the `negative` to be `positive` as Danil suggested for the others. Is zero allowed for positive?

Comment (jmalloc) [reply config/int.go]: Nah, `0` is not allowed - I have fixed these.

**Review by danilvpetrov — COMMENTED** [review]
Why is zero not allowed?

**Comment by jmalloc** [comment]
> Why is zero not allowed?

As discussed on slack, I've replaced the `GetIntP()` variants with `GetIntBetween()`.

**Comment by jmalloc** [comment]
Ok, all done I'm pretty sure.

**Review by danilvpetrov — APPROVED** [review]
Looked through the entire code. I definitely like the separation between `AsIntXXX()` and `AsUIntXXX()` functions. Looks awesome!

File: config/float32_test.go [hunk 9]
```diff
@@ -1,201 +1,153 @@
...
 	})
 
 	It("returns the default value if the key is not defined", func() {
 		b := Map{}
 
-		v, err := GetFloat32Default(b, "<key>", 456.78)
-		Expect(err).ShouldNot(HaveOccurred())
-		Expect(v).To(BeNumerically("~", float32(456.78)))
+		v := AsFloat32Default(b, "<key>", 50)
+		Expect(v).To(BeNumerically("==", 50))
```
Comment (koden-km) [inline config/float32_test.go:46]: Should these float tests use float looking values, `50.0`?

File: config/float64_test.go [hunk 10]
```diff
@@ -6,138 +6,148 @@ import (
...
+			AsFloat64Between(b, "<key>", -100, 100)
+		}).To(PanicWith(`expected <key> to be a 64-bit floating-point number: strconv.ParseFloat: parsing "<invalid>": invalid syntax`))
 	})
 })
 
-var _ = Describe("func MustGetFloat64Default()", func() {
-	It("returns a positive float value", func() {
-		b := Map{"<key>": String("123.45")}
+var _ = Describe("func AsFloat64DefaultBetween()", func() {
+	It("returns an float64 value", func() {
```
Comment (koden-km) [inline config/float64_test.go:100]: Is this correct grammar? Should it be "returns a float64 value"?
> returns an float64 value

File: config/uint.go [hunk 11]
```diff
@@ -2,95 +2,236 @@ package config
...
 
-// GetUint returns the uint representation of the value associated with k.
+const (
+	// MaxUint is the maximum value that can be expressed using the uint type.
+	MaxUint = 1<<uintBitSize - 1 // 1<<32 - 1 or 1<<64 - 1
+
+	uintBitSize = 32 << (^uint(0) >> 32 & 1) // 32 or 64
+)
+
+// AsUint returns the int representation of the value associated with k or panics
```
Comment (koden-km) [inline config/uint.go:16]: Should this say "unsigned int"?
> // AsUint returns the int representation

File: config/uint.go [hunk 12]
```diff
@@ -2,95 +2,236 @@ package config
...
+// AsUintDefaultBetween returns the int representation of the value associated
+// with k, or the default value v if k is undefined.
 //
-// If k is undefined, ok is false and err is nil.
+// It panics if the value is not between min and max (inclusive).
+func AsUintDefaultBetween(b Bucket, k string, v, min, max int) uint {
+	return uint(asUintDefault(b, k, 0, uint64(v), uint64(min), uint64(max)))
+}
+
+// AsUint8 returns the int8 representation of the value associated with k or
```
Comment (koden-km) [inline config/uint.go:44]: All these specific sized **uint** doc blocks below just say `int#` instead of `uint#`?

**Review by koden-km — APPROVED** [review]

File: config/float64_test.go (hunk 10 above)
Comment (jmalloc) [reply config/float64_test.go]: Fixed.

File: config/uint.go (hunk 12 above)
Comment (jmalloc) [reply config/uint.go]: Fixed.

File: config/float32_test.go (hunk 9 above)
Comment (jmalloc) [reply config/float32_test.go:46]: Yep, will fix.

**Comment by jmalloc** [comment]
Thanks again guys.

**Pull request merged**